dyn_load.c

```diff
@@ -480,18 +480,56 @@ STATIC int GC_register_dynlib_callback(struct dl_phdr_info * info,
 
   p = info->dlpi_phdr;
   for( i = 0; i < (int)info->dlpi_phnum; i++, p++ ) {
-    switch( p->p_type ) {
+    if( p->p_type == PT_LOAD) {
+      GC_has_static_roots_func callback = GC_has_static_roots;
+      if( !(p->p_flags & PF_W) ) continue;
+      start = ((char *)(p->p_vaddr)) + info->dlpi_addr;
+      end = start + p->p_memsz;
+
+      if (callback != 0 && !callback(info->dlpi_name, start, p->p_memsz))
+        continue;
+      
 #     ifdef PT_GNU_RELRO
-        case PT_GNU_RELRO:
+#       if CPP_WORDSZ == 64
+          /* FIXME: GC_push_all eventually does the correct         */
+          /* rounding to the next multiple of ALIGNMENT, so, most   */
+          /* probably, we should remove the corresponding assertion */
+          /* check in GC_add_roots_inner along with this code line. */
+          /* start pointer value may require aligning */
+          start = (ptr_t)((word)start & ~(sizeof(word) - 1));
+#       endif
+        if (n_load_segs >= MAX_LOAD_SEGS) {
+          WARN("Too many PT_LOAD segs\n", 0);
+          GC_add_roots_inner(start, end, TRUE);
```

Comment:
I'm not sure what you think about, but I would think, with my limited knowledge about bdwgc, that all roots added with GC_add_roots_inner are cleared after collecting. So I guess things could go really wrong if a library is unloaded while collecting garbage, but not if it is unloaded inbetween two collections.